src/data/finsupp/ne_locus.lean

```diff
@@ -0,0 +1,114 @@
+`finsupp.support (f - g)`.
+-/
+
+variables {α N : Type*}
+
+namespace finsupp
+variables [decidable_eq α] [decidable_eq N]
+
+section ne_locus
+
+section N_has_zero
+variables [has_zero N] {f g : α →₀ N}
+
+/--  Given two finitely supported functions `f g : α →₀ N`, `finsupp.ne_locus f g` is the `finset`
+where `f` and `g` differ. This generalizes `(f - g).support` to situations without subtraction. -/
+def ne_locus (f g : α →₀ N) : finset α :=
+(f.support ∪ g.support).filter (λ x, f x ≠ g x)
+
+@[simp]
+lemma mem_ne_locus {a : α} : a ∈ f.ne_locus g ↔ f a ≠ g a :=
+by simpa only [ne_locus, finset.mem_filter, finset.mem_union, mem_support_iff,
+    and_iff_right_iff_imp] using ne.ne_or_ne _
+
+lemma coe_ne_locus : ↑(f.ne_locus g) = {x | f x ≠ g x} :=
+by { ext, exact mem_ne_locus }
+
+@[simp] lemma ne_locus_eq_empty : f.ne_locus g = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp [ne_locus]⟩,
+  ext a,
+  exact not_not.mp (mem_ne_locus.not.mp (finset.eq_empty_iff_forall_not_mem.mp h a)),
+end
+
+@[simp] lemma nonempty_ne_locus_iff : (f.ne_locus g).nonempty ↔ f ≠ g :=
+finset.nonempty_iff_ne_empty.trans ne_locus_eq_empty.not
+
+variables (f g)
+
+lemma ne_locus_comm (f g : α →₀ N) : f.ne_locus g = g.ne_locus f :=
+by simp_rw [ne_locus, finset.union_comm, ne_comm]
+
+@[simp]
+lemma ne_locus_zero_right : f.ne_locus 0 = f.support :=
+by { ext, rw [mem_ne_locus, mem_support_iff, coe_zero, pi.zero_apply] }
+
+@[simp]
+lemma ne_locus_zero_left : (0 : α →₀ N).ne_locus f = f.support :=
+(ne_locus_comm _ _).trans (ne_locus_zero_right _)
+
+lemma subset_map_range_ne_locus {M} [decidable_eq M] [has_zero M] {F : N → M} (F0 : F 0 = 0) :
+  (f.map_range F F0).ne_locus (g.map_range F F0) ⊆ f.ne_locus g :=
+begin
+  refine λ x, _,
+  simp only [mem_ne_locus, map_range_apply, not_imp_not],
+  exact congr_arg _,
+end
+
+lemma map_range_ne_locus_eq {M} [decidable_eq M] [has_zero M]
+  {F : N → M} (F0 : F 0 = 0) (hF : function.injective F) :
+  (f.map_range F F0).ne_locus (g.map_range F F0) = f.ne_locus g :=
+by { ext, simpa only [mem_ne_locus] using hF.ne_iff }
+
+end N_has_zero
+
+lemma add_ne_locus_add_eq_left [add_left_cancel_monoid N] (f g h : α →₀ N) :
+  (f + g).ne_locus (f + h) = g.ne_locus h  :=
+begin
+  ext,
+  simp only [ne_locus, ne.def, add_right_inj, finset.mem_filter, finset.mem_union, mem_support_iff,
+    coe_add, pi.add_apply, and.congr_left_iff],
+  exact λ bc, ⟨λ h, ne.ne_or_ne 0 bc, λ h, ne.ne_or_ne _ ((add_right_inj _).not.mpr bc)⟩,
+end
+
+--  can this proof by replaced by the previous one applied to `Nᵃᵒᵖ` and interlaced with `op unop`?
+lemma add_ne_locus_add_eq_right [add_right_cancel_monoid N] (f g h : α →₀ N) :
+  (f + h).ne_locus (g + h) = f.ne_locus g  :=
+begin
+  ext,
+  simp only [ne_locus, ne.def, add_left_inj, finset.mem_filter, finset.mem_union, mem_support_iff,
+    coe_add, pi.add_apply, and.congr_left_iff],
+  exact λ bc, ⟨λ h, ne.ne_or_ne 0 bc, λ h, ne.ne_or_ne _ ((add_left_inj _).not.mpr bc)⟩,
+end
```

Comment:
Golfs to
```suggestion
map_range_ne_locus_eq _ _ _ _ (λ gn, add_left_injective _)
```
with
```lean
lemma map_range_ne_locus_eq {F : M → N → P} (F0 : F 0 0 = 0)
  (f₁ f₂ : α →₀ M) (g : α →₀ N) (hF : ∀ g, function.injective (λ f, F f g)) :
  (zip_with F F0 f₁ g).ne_locus (zip_with F F0 f₂ g) = f₁.ne_locus f₂ :=
by { ext, simpa only [mem_ne_locus] using (hF _).ne_iff }
```

Reply:
I have not fully parsed your version of `map_range_ne_locus_eq`.  I imagine that the version that I had is a special case, but I have not been able to find the correct specialization.  I left both, the old one with a `'`, but I will try to remove the primed version, once I figure out how to!